embedded/appendable/remoteapp/remote_app.go

```diff
@@ -324,36 +310,32 @@ func (r *RemoteStorageAppendable) downloadChunk(chunkID int64) {
 		var flTmp *os.File
 		defer func() { flTmp.Close() }()
 
-		cp := r.chunkedProcess(ctx)
-		cp.RetryableStep(func(retries int, delay time.Duration) (retryNeeded bool, err error) {
-
+		cp := chain.New().WithContext(ctx).WithBackoff(r.backoff)
+		cp.NextWithRetry(func() (err error) {
 			data, err := r.rStorage.Get(ctx, r.remotePath+r.appendableName(chunkID), 0, -1)
 			if err != nil {
 				metricsDownloadRetried.Inc()
-				return true, nil
+				return nil
 			}
 			defer data.Close()
 
 			flTmp, err = os.OpenFile(fileNameTmp, os.O_CREATE|os.O_RDWR, r.fileMode)
 			if err != nil {
 				// Couldn't create temporary local file, something is broken on local FS
-				return false, err
+				return err
```

Comment:
This is different behavior from the previous implementation.

In case the local temporary file can not be created, we don't even want to retry this. Instead the whole chain should fail with an error without retries.

I think that currently there's no way to break retrying - any error returned will cause retry, correct? And returning nil will continue with the next step.

Reply:
Maybe we should experiment with wrapping errors?

So in order to retry when error happens e.g.:

```go
if err != nil {
  if error.Is(err, ErrThatCanBeRetried) {
    // Will retry after backoff timeout
    return retry.WithError(err)
  }
  // Will not retry and fail the whole chain
  return err
}

if shouldRetryAnyway {
  // Will retry after backoff timeout
  return retry.WithoutError()
}

// This will not retry and will continue with the following step
return nil
```

The `retry.WithError` and `retry.WithoutError` could then return a dedicated error declared in the `retry` package, it could also be extended in the future with things like whether we want to skip the backoff sleep or set the maximum number of retries allowed etc.:

```go
return retry.WithError(err).SkipBackoffSleep()

return retry.WithoutError().MaxRetries(10)
```

^ that I would keep for future updates to avoid adding code that we don't currently use.

Reply:
Actually the `Next` step will return error and block other steps without retrying, I've kept the functionality as close as possible to chunkedProcess, but I'll have a look at this

Reply:
I've just added a different step to the chain function, and added a retriable function, where we can specify which error to retry on, eg:

```go
		cp := chain.New().WithContext(ctx).WithBackoff(r.backoff).WithRetriable(func(err error) bool {
			switch {
			case err == retry.ErrNoRetry:
				return false
			default:
				return true
			}
		})
		cp.NextWithRetry(func() (err error) {
			data, err := r.rStorage.Get(ctx, r.remotePath+r.appendableName(chunkID), 0, -1)
			if err != nil {
				metricsDownloadRetried.Inc()
				return err
			}
			defer data.Close()

			flTmp, err = os.OpenFile(fileNameTmp, os.O_CREATE|os.O_RDWR, r.fileMode)
			if err != nil {
				// Couldn't create temporary local file, something is broken on local FS
				return retry.ErrNoRetry
			}

			_, err = io.Copy(flTmp, data)
			if err != nil {
				metricsDownloadRetried.Inc()
				return nil
			}
			return nil
		})
```